src/math/lp/lar_solver.cpp

```diff
@@ -312,38 +312,87 @@ namespace lp {
         }
     }
 
-    bool lar_solver::improve_bound(lpvar j, bool improve_lower_bound) {
+    // get dependencies of the corresponding bounds from max_coeffs
+    u_dependency* lar_solver::get_dependencies_of_maximum(const vector<std::pair<mpq,lpvar>>& max_coeffs) {        
+        const auto& s = this->m_mpq_lar_core_solver.m_r_solver;
+        // The linear combinations of p.first*p.second = the term that got maximized
+        // Every j with positive coeff is at its upper bound,
+        // and every j with negative coeff is at its lower bound: so the sum cannot be increased.
+        // All variables j in the sum are non-basic.
+        u_dependency* dep = nullptr;
+        for (const auto & p: max_coeffs) {
+            const auto & d_j = p.first;
+            unsigned j = p.second;
+            SASSERT (!d_j.is_zero());
+                
+            TRACE("lar_solver_improve_bounds", tout << "d[" << j << "] = " << d_j << "\n";
+                                               s.print_column_info(j, tout););
+            const ul_pair& ul = m_columns_to_ul_pairs[j];
+            u_dependency * bound_dep;
+            if (d_j.is_pos()) {
+                bound_dep = ul.upper_bound_witness();
+            } 
+            else {
+                bound_dep = ul.lower_bound_witness();               
+            }
+            TRACE("lar_solver_improve_bounds", {
+	           svector<constraint_index> cs;
+               m_dependencies.linearize(bound_dep, cs);
+               for (auto c : cs) 
+                    m_constraints.display(tout, c) << "\n";
+            });
+            dep = m_dependencies.mk_join(dep, bound_dep);
+        }
+        return dep;
+    }
+    // returns nullptr if the bound is not improved, otherwise returns the witness of the bound
+    u_dependency* lar_solver::find_improved_bound(lpvar j, bool lower_bound, mpq& bound) {
+        SASSERT(is_feasible());
+        if (lower_bound) {
+            if (column_has_lower_bound(j) && get_column_value(j) == column_lower_bound(j))
+                return nullptr;  // cannot do better
+        }
+        else {  // improve upper bound = false
+            if (column_has_upper_bound(j) && get_column_value(j) == column_upper_bound(j))
+                return nullptr;  // cannot do better
+        }
+
         lar_term term = get_term_to_maximize(j);
-        if (improve_lower_bound)
+        if (lower_bound)
             term.negate();
-        impq bound;
-        if (!maximize_term_on_corrected_r_solver(term, bound))
-            return false;
-        
-        return false;
-        // TODO
-        if (improve_lower_bound) {
+        vector<std::pair<mpq, unsigned>> max_coeffs;
+        TRACE("lar_solver_improve_bounds", tout << "j = " << j << ", "; print_term(term, tout << "term to maximize\n"););
+        impq term_max;
+        if (!maximize_term_on_feasible_r_solver(term, term_max, &max_coeffs))
+            return nullptr;
+        bound = term_max.x;
```

Comment:
add comment that strictness gets ignored (bounds tightening is a heuristic anyhow)